### PR TITLE
f-http@v0.2.0 - Developing POC

### PR DESCRIPTION
**Pull request by tommcclean**
- This PR develops the POC in the following ways - [checkout the readme if you want to feedback on how it works](https://github.com/tommcclean/fozzie-components/blob/f-http%40v0.2.0/packages/services/f-http/README.md) so far
- Its not finished yet, just trying to use checkpoints so I don't end up with a big PR

### Added
- Support for configuration merging
- Support for basic GET and POST requests

### Changed
- Updated Readme documentation

### CODE REVIEW

File: packages/services/f-http/src/createClient.js [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+/**
+    * Get a resource
+    * @param {string} resource - The resource to post (URL)
+    * @return {object} - Returns data from response
+*/
+const get = async resource => {
+    try {
+        const response = await _axiosInstance.get(resource);
+
+        return response.data;
```
Comment (BanksySan) [inline packages/services/f-http/src/createClient.js:22]: Potential null exception if the response isn't a 2xx.

File: packages/services/f-http/src/createClient.js [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    } catch (error) {
+        if (_configuration.errorCallback) {
+            _configuration.errorCallback(error, resource);
```
Comment (BanksySan) [inline packages/services/f-http/src/createClient.js]: We should wrap this in a `try-catch`.

File: packages/services/f-http/src/createClient.js [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        const config = {
+            headers
+        };
+
+        const response = await _axiosInstance.post(resource, body, config);
+
+        return response.data;
+    } catch (error) {
+        if (_configuration.errorCallback) {
+            _configuration.errorCallback(error, resource);
```
Comment (BanksySan) [inline packages/services/f-http/src/createClient.js]: Same here, catch exception.

File: packages/services/f-http/src/createClient.js [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+    _axiosInstance = axios.create({
+        baseURL: _configuration.baseUrl,
+        timeout: _configuration.timeout
+    });
+
+    return {
+        get,
+        post,
+        readConfiguration: () => _configuration
```
Comment (BanksySan) [inline packages/services/f-http/src/createClient.js:69]: Why is this a method rather than a variable?

File: packages/services/f-http/src/tests/index.test.js [hunk 5]
```diff
@@ -1,11 +1,22 @@
...
+
+    it('httpClient should expose expected methods', async () => {
+        // Arrange & Act
+        const httpClient = createClient();
+
+        // Assert
+        expect(httpClient).toBeDefined();
+        expect(httpClient.get).toBeDefined();
+        expect(httpClient.post).toBeDefined();
+        expect(httpClient.readConfiguration).toBeDefined();
```
Comment (BanksySan) [inline packages/services/f-http/src/tests/index.test.js:19]: Should there be test cases for null being returned for `data` and for the calls to `errorCallback` throwing?

**Review by BanksySan — DISMISSED** [review]

File: packages/services/f-http/src/createClient.js (hunk 1 above)
Comment (tommcclean) [reply packages/services/f-http/src/createClient.js:22]: Would be handled by the try-catch and throw to the caller who should handle it, any bad response code would fail before this line as well. Additionally the caller can even provide a function to globally handle errors which this would fall into or they could opt to do as most do already and wrap a try catch around their call.

File: packages/services/f-http/src/createClient.js (hunk 2 above)
Comment (tommcclean) [reply packages/services/f-http/src/createClient.js]: Its in a try catch, this is a call-back provided by the caller; so should their own callback fail they are responsible for handling that by handling errors in their own code. They are forced to consider that, as the default is no callback; and thus would simply rethrow to the line where it is called; rather than the installation code. So its kind of an optional way of handling it all in one place.

File: packages/services/f-http/src/createClient.js (hunk 3 above)
Comment (tommcclean) [reply packages/services/f-http/src/createClient.js]: Same as above

File: packages/services/f-http/src/tests/index.test.js (hunk 5 above)
Comment (tommcclean) [reply packages/services/f-http/src/tests/index.test.js:19]: I intend to put some tests in for the GET, PUT, POST, PATCH, DELETE's when they are all in there as part of the same ticket; just trying to keep the PR's manageable and small. Much of that code will be shared between them and thus the tests too.

File: packages/services/f-http/src/createClient.js (hunk 4 above)
Comment (tommcclean) [reply packages/services/f-http/src/createClient.js:69]: Just a matter of preference, makes no difference in reality; though it may be desirable to return a new version of the object from the function rather than exposing the reference later.

File: packages/services/f-http/src/createClient.js [hunk 6]
```diff
@@ -0,0 +1,73 @@
+import axios from 'axios';
+import defaultOptions from './defaultOptions';
+
+let _configuration = null;
+let _axiosInstance = null;
+
+/**
+    * Get a resource
+    * @param {string} resource - The resource to post (URL)
```
Comment (SachaZvetelman) [inline packages/services/f-http/src/createClient.js]: "The resource to get", right? 😬

File: packages/services/f-http/src/createClient.js [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+    * @return {object} - Returns data from response
+*/
+const get = async resource => {
```
Comment (SachaZvetelman) [inline packages/services/f-http/src/createClient.js]: I think we should have the ability to pass headers in here as well.

File: packages/services/f-http/src/createClient.js [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+        const response = await _axiosInstance.post(resource, body, config);
+
+        return response.data;
+    } catch (error) {
+        if (_configuration.errorCallback) {
+            _configuration.errorCallback(error, resource);
+        }
+
+        throw error;
```
Comment (SachaZvetelman) [inline packages/services/f-http/src/createClient.js]: It seems that this snippet will be repeated pretty much across all the different request types, so going forward you might benefit from extracting this into a separate function.

Comment (tommcclean) [reply packages/services/f-http/src/createClient.js:47]: Thats the plan

File: packages/services/f-http/src/createClient.js (hunk 6 above)
Comment (tommcclean) [reply packages/services/f-http/src/createClient.js]: Fixed

File: packages/services/f-http/src/createClient.js (hunk 7 above)
Comment (tommcclean) [reply packages/services/f-http/src/createClient.js]: Fair - Implemented

File: packages/services/f-http/src/createClient.js (hunk 8 above)
Comment (tommcclean) [reply packages/services/f-http/src/createClient.js]: Extracted and tested

File: packages/services/f-http/src/tests/errorHandler.test.js [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+        const errorCallbackMock = jest.fn();
+
+        // Act
+        handleError(error, errorCallbackMock);
+
+        // Assert
+        expect(errorCallbackMock.mock.calls.length).toBe(1);
+
+        // Check that the first argument is the error
+        expect(errorCallbackMock.mock.calls[0][0]).toBe(error);
```
Comment (SachaZvetelman) [inline packages/services/f-http/src/tests/errorHandler.test.js]: Would `expect(errorCallbackMock).toHaveBeenCalledWith(error);` work?

Comment (tommcclean) [reply packages/services/f-http/src/tests/errorHandler.test.js]: Ta, much better

File: packages/services/f-http/src/tests/errorHandler.test.js [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+            message: 'This is an example error'
+        };
+
+        const errorCallbackMock = jest.fn();
+
+        // Act
+        handleError(error, errorCallbackMock);
+
+        // Assert
+        expect(errorCallbackMock.mock.calls.length).toBe(1);
```
Comment (SachaZvetelman) [inline packages/services/f-http/src/tests/errorHandler.test.js]: You can get rid of this now.

File: packages/services/f-http/src/tests/errorHandler.test.js [hunk 11]
```diff
@@ -0,0 +1,38 @@
+import handleError from '../errorHandler';
+
+describe('errorHandler', () => {
+    it('errorHandler should be defined', () => {
```
Comment (SachaZvetelman) [inline packages/services/f-http/src/tests/errorHandler.test.js]: No need to mention `errorHandler` in every test name since their `describe` block already mentions it.

File: packages/services/f-http/src/tests/errorHandler.test.js [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+    });
+
+    it('errorHandler should re-throw error when callback is not provided', () => {
+        // Arrange
+        const expectedError = new Error('This is an example error');
+
+        try {
+            // Act
+            handleError(expectedError);
+            throw new Error('Test Failed');
```
Comment (SachaZvetelman) [inline packages/services/f-http/src/tests/errorHandler.test.js]: Doesn't this defeat the purpose of the test? Shouldn't this throw be done by `handleError`?

File: packages/services/f-http/src/tests/errorHandler.test.js (hunk 10 above)
Comment (tommcclean) [reply packages/services/f-http/src/tests/errorHandler.test.js]: I guess I still want to make sure it is just called once and its clear from the test that is the case

Comment (SachaZvetelman) [reply packages/services/f-http/src/tests/errorHandler.test.js]: I might be wrong, but I think the `toHaveBeenCalledWith();` will fail if it's called more than once. If I'm wrong, you can replace your line with `expect(errorCallbackMock).toHaveBeenCalledTimes(1);` which in my opinion looks slightly nicer and more readable.

Comment (tommcclean) [reply packages/services/f-http/src/tests/errorHandler.test.js]: Ill go with the latter; prefer it to be obvious if not from the way the test functions, but how it reads.

**Review by SachaZvetelman — APPROVED** [review]
Looking like a great and promising package! 📦 🚀

File: packages/services/f-http/src/tests/createClient.test.js [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+            expect(createClient).toBeDefined();
+        });
+
+        it('should use default options when not overridden', async () => {
+            // Arrange & Act
+            const httpClient = createClient();
+            const mergedOptions = httpClient.readConfiguration();
+
+            // Assert
+            expect(mergedOptions).toBeDefined();
```
Comment (kevinrodrigues) [inline packages/services/f-http/src/tests/createClient.test.js:16]: You could shorten this to just checking the object itself? That way you will just have one expect?

File: packages/services/f-http/src/tests/createClient.test.js [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+
+            const httpClient = createClient({
+                baseUrl: expectedBaseUrl,
+                timeout: expectedTimeout,
+                errorCallback: expectedErrorCallback
+            });
+            const mergedOptions = httpClient.readConfiguration();
+
+            // Assert
+            expect(mergedOptions).toBeDefined();
```
Comment (kevinrodrigues) [inline packages/services/f-http/src/tests/createClient.test.js:36]: Same.

**Review by kevinrodrigues — APPROVED** [review]

File: packages/services/f-http/src/tests/createClient.test.js (hunk 13 above)
Comment (tommcclean) [reply packages/services/f-http/src/tests/createClient.test.js:16]: Yeah good idea, ill switch to this in the next version
